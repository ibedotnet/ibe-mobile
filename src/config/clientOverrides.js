const clientOverrides = {
  1331: { forceDisableSwitch: true }, // Switch is always disabled for this client
  // Other clients are not listed, meaning they follow the default behavior
};

export default clientOverrides;