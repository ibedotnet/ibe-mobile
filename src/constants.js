/**
 * Function to dynamically determine the current environment.
 * This function can be based on various factors such as the domain, configuration files, etc.
 * @returns {string} - The current environment (e.g., "development", "testing", "production").
 */
const determineEnvironment = () => {
  return "testing"; // Default environment for demonstration purposes
};

/**
 * Enum representing different environments with their base URLs.
 * @readonly
 * @enum {string}
 */
const Environment = {
  /** Development environment. */
  DEVELOPMENT: "development",
  /** Testing environment. */
  TESTING: "testing",
  /** Production environment. */
  PRODUCTION: "production",
};

/**
 * Base URLs for different environments.
 * @constant
 * @type {Object.<string, string>}
 */
const BASE_URL = {
  [Environment.DEVELOPMENT]: "http://172.17.0.80",
  [Environment.TESTING]: "http://testserver.ibenv.net",
  [Environment.PRODUCTION]: "http://appnew.ibenv.net",
};

// Determine the current environment
const currentEnvironment = determineEnvironment();

/**
 * Base URL for the current environment.
 * @constant
 * @type {string}
 */
const baseURL = BASE_URL[currentEnvironment];

/**
 * Object containing API endpoints based on the current environment's base URL.
 * @constant
 * @type {Object.<string, string>}
 */
const API_ENDPOINTS = {
  /** Endpoint for authentication. */
  AUTHENTICATE: `${baseURL}/api/authenticate`,
  /** Endpoint for user logout. */
  LOGOUT: `${baseURL}/api/logout`,
  /** Endpoint for querying fields. */
  QUERY: `${baseURL}/api/queryfields`,
  /** Endpoint for updating fields. */
  UPDATE_FIELDS: `${baseURL}/api/updatefields`,
  /** Endpoint for accessing resources. */
  RESOURCE: `${baseURL}/resource`,
  /** Endpoint for obtaining a new object ID. */
  NEW_OBJECT_ID: `${baseURL}/api/newobjectid`,
  /** Endpoint for uploading binary resources. */
  UPLOAD_BINARY_RESOURCE: `${baseURL}/api/uploadbinaryresource`,
  /** Endpoint for invoking APIs. */
  INVOKE: `${baseURL}/api/invoke`,
  /** Endpoint for invoking new APIs. */
  INVOKE_API_NEW: `${baseURL}/api/invokeApiNew`,
  /** Endpoint for invoking get document workflow status. */
  GET_DOCUMENT_STATUS: `${baseURL}/api/getdocstatus`,
  /** Endpoint for invoking set document workflow status. */
  SET_DOCUMENT_STATUS: `${baseURL}/api/setdocstatus`,
};

/**
 * Timeout duration for API requests in milliseconds.
 * @constant
 * @type {number}
 */
const API_TIMEOUT = 120000; // 2 minutes

/**
 * Function to dynamically determine the application version.
 * This function could involve reading the version from a file, fetching it from a server, etc.
 * @returns {string} - The application version.
 */
const getVersion = () => {
  return "1.0.0"; // Default version for demonstration purposes
};

/**
 * Object containing application-related constants.
 * @constant
 * @type {Object}
 */
const APP = {
  /** The application version. */
  VERSION: getVersion(),
  /** Client ID of the logged-in user. */
  LOGIN_USER_CLIENT: "",
  /** Date format for the logged-in user. */
  LOGIN_USER_DATE_FORMAT: "",
  /** Employee ID of the logged-in user. */
  LOGIN_USER_EMPLOYEE_ID: "",
  /** ID of the logged-in user. */
  LOGIN_USER_ID: "",
  /** Language preference of the logged-in user. Default is English. */
  LOGIN_USER_LANGUAGE: "en",
  /** Work schedule external name of the logged-in user.*/
  LOGIN_USER_WORKSCHEDULE_NAME: "",
  /** Timesheet type of the logged-in user work schedule.*/
  LOGIN_USER_WORK_SCHEDULE_TIMESHEET_TYPE: "",
};

/**
 * Object containing activity IDs for different application modules.
 * @constant
 * @type {Object}
 */
const APP_ACTIVITY_ID = {
  /** Activity ID for timesheet module. */
  TIMESHEET: "manage.timemgt.timesheet",
  /** Activity ID for expense module. */
  EXPENSE: "manage.expensemgt.expenseclaim",
  /** Activity ID for absence module. */
  ABSENCE: "manage.absencemgt.myabsencerequest",
  /** Activity ID for inbox module. */
  INBOX: "platform.bpm.inbox",
};

/**
 * Object containing app path names for different application modules.
 * @constant
 * @type {Object}
 */
const APP_NAME = {
  /** App name for timesheet module. */
  TIMESHEET: "hrAdmin.timemgt",
  /** App name for expense module. */
  EXPENSE: "accountspayable.expenseclaim",
  /** App name for absence module. */
  ABSENCE: "hrAdmin.absencemgmt",
  /** App name for employee module. */
  EMPLOYEE: "hrAdmin.employeemaster",
};

/**
 * Object containing business object categories and their mappings.
 * @constant
 * @type {Object}
 */
const BUSOBJCAT = {
  /** Business object category for timesheet. */
  TIMESHEET: "Timesheet",
  /** Business object category for expense. */
  EXPENSE: "Expense",
  /** Business object category for absence. */
  ABSENCE: "Absence",
  /** Business object category for inbox. */
  INBOX: "Inbox",
};

/**
 * Object mapping business object categories to their respective data models.
 * @constant
 * @type {Object}
 */
const BUSOBJCATMAP = {
  [BUSOBJCAT.TIMESHEET]: "TimeConfirmation",
  [BUSOBJCAT.EXPENSE]: "ExpenseClaim",
  [BUSOBJCAT.ABSENCE]: "Absence",
};

/**
 * Time duration (in milliseconds) representing the interval between consecutive clicks
 * in a double-click event.
 * @constant
 * @type {number}
 */
const DOUBLE_CLICK_DELTA = 400;

/**
 * Object containing internal status codes.
 * @constant
 * @enum {number}
 */
const INTSTATUS = {
  /** Indicates an active status. */
  ACTIVE: 0,
  /** Indicates a deleted status. */
  DELETED: 3,
};

/**
 * Maximum lengths for login input fields.
 * @constant
 * @type {Object}
 */
const LOGIN_INPUTS_MAXLENGTH = {
  /** Maximum length for username field. */
  USERNAME: 10,
  /** Maximum length for password field. */
  PASSWORD: 20,
  /** Maximum length for client ID field. */
  CLIENTID: 4,
};

/**
 * Maximum allowed upload file size in MB.
 * @constant
 * @type {number}
 */
const MAX_UPLOAD_FILE_SIZE = 10;

/**
 * Default page size for paginated data.
 * @constant
 * @type {number}
 */
const PAGE_SIZE = 20;

/**
 * Flag indicating whether the application is running in test mode.
 * @constant
 * @type {boolean}
 */
const TEST_MODE = false;

/**
 * Array containing valid file extensions.
 * These extensions are commonly used for text documents, images, and other file types.
 * @constant
 * @type {string[]}
 */
const VALID_FILE_EXTENSIONS = [
  ".txt",
  ".docx",
  ".pdf",
  ".jpg",
  ".png",
  ".xlsx",
  ".xls",
  ".mp4",
  ".jpeg",
  ".zip",
  ".html",
  ".mp3",
  ".pptx",
  ".csv",
  ".json",
  ".xml",
  ".svg",
  ".wav",
  ".avi",
];

export {
  API_ENDPOINTS,
  API_TIMEOUT,
  APP,
  APP_ACTIVITY_ID,
  APP_NAME,
  BUSOBJCAT,
  BUSOBJCATMAP,
  DOUBLE_CLICK_DELTA,
  INTSTATUS,
  LOGIN_INPUTS_MAXLENGTH,
  MAX_UPLOAD_FILE_SIZE,
  PAGE_SIZE,
  TEST_MODE,
  VALID_FILE_EXTENSIONS,
};
