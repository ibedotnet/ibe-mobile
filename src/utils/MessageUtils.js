import Toast from "react-native-root-toast";

// Internal queue to manage toast messages
let toastQueue = [];
let isToastVisible = false;

/**
 * Displays a toast message using the React Native Root Toast library.
 * Toast messages are queued to ensure they are displayed sequentially.
 *
 * @param {string} message - The message to be displayed in the toast.
 * @param {string} type - The type of the toast message. Possible values are 'information', 'warning', or 'error'. Default is 'information'.
 * @param {object} props - Additional props to be passed to the Toast component.
 */
const showToast = (message, type = "information", props) => {
  if (!Toast) {
    console.warn("Toast library not properly initialized.");
    return;
  }

  // Check if the message is already in the queue
  if (toastQueue.some((item) => item.message === message)) {
    return;
  }

  // Add the new message to the queue
  toastQueue.push({ message, type, props });

  // If no toast is currently visible, display the first message in the queue
  if (!isToastVisible) {
    displayNextToast();
  }
};

/**
 * Displays the next toast message in the queue.
 * Toast messages are displayed sequentially and removed from the queue after being shown.
 */
const displayNextToast = () => {
  if (toastQueue.length > 0) {
    const { message, type, props } = toastQueue[0];
    let duration = Toast.durations.SHORT; // Default duration

    // Determine background color and text color based on the type of message
    let backgroundColor, textColor;
    if (type === "information") {
      backgroundColor = "#007bff"; // Bootstrap blue
      textColor = "#fff";
    } else if (type === "warning") {
      backgroundColor = "#ffc107"; // Bootstrap yellow
      textColor = "#000";
      duration = Toast.durations.LONG; // Longer duration for warnings
    } else if (type === "error") {
      backgroundColor = "#dc3545"; // Bootstrap red
      textColor = "#fff";
      duration = Toast.durations.LONG; // Longer duration for errors
    }

    // Display the toast
    Toast.show(message, {
      backgroundColor,
      textColor,
      duration,
      ...props,
      onHidden: () => {
        // Remove the displayed message from the queue
        toastQueue.shift();

        // Set flag to indicate that no toast is currently visible
        isToastVisible = false;

        // Display the next message in the queue
        displayNextToast();
      },
    });

    // Set flag to indicate that a toast is currently visible
    isToastVisible = true;
  }
};

export { showToast };
