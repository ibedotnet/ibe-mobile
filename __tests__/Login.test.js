// Third-party libraries
import "@testing-library/jest-native/extend-expect";
import React from "react";
import {
  render,
  screen,
  fireEvent,
  waitFor,
} from "@testing-library/react-native";
import {
  mockNavigation,
  MockNavigationContainer,
} from "./mock/mockNavigation.js";

// Setup and mocks
import "./config/setupTests.js";
import { MockLoggedInUserInfoProvider } from "./mock/MockLoggedInUserInfoProvider.js";
import { translations } from "./config/testTranslations"; // Import translations

// App components and utilities
import { API_ENDPOINTS } from "../src/constants.js";
import Login from "../src/screens/Login.js";
import Home from "../src/screens/Home.js";
import { fetchData } from "../src/utils/APIUtils";
import { showToast } from "../src/utils/MessageUtils";

import * as SecureStore from "expo-secure-store";

describe("Login Component", () => {
  beforeEach(() => {
    jest.clearAllMocks(); // Clear previous mocks before each test
  });

  // Utility function to render Login component with navigation
  const renderLogin = () => {
    return render(
      <MockNavigationContainer>
        <MockLoggedInUserInfoProvider>
          <Login navigation={mockNavigation} />
        </MockLoggedInUserInfoProvider>
      </MockNavigationContainer>
    );
  };

  /**
   * Test Case 1: Renders the Login component correctly
   * Ensures that the Login component renders all expected input fields and buttons.
   */
  it("should render the Login component correctly", () => {
    // Render the Login component
    renderLogin();

    // Check that input fields are rendered
    const usernameInput = screen.getByPlaceholderText(
      translations.login_username_placeholder
    );
    const passwordInput = screen.getByPlaceholderText(
      translations.login_password_placeholder
    );
    const clientIdInput = screen.getByPlaceholderText(
      translations.login_clientId_placeholder
    );

    expect(usernameInput).toBeTruthy();
    expect(passwordInput).toBeTruthy();
    expect(clientIdInput).toBeTruthy();

    // Check that the login button is rendered
    const loginButton = screen.getByRole("button", {
      name: translations.login_button_text,
    });
    expect(loginButton).toBeTruthy();

    // Optionally, check that the "Remember Me" checkbox is rendered
    const rememberMeCheckbox = screen.getByRole("checkbox", {
      name: translations.login_rememberMe_text,
    });
    expect(rememberMeCheckbox).toBeTruthy();

    // Check that the password visibility toggle button is rendered
    const toggleButton = screen.getByLabelText(translations.show_password);
    expect(toggleButton).toBeTruthy();

    // Check for the version text containing "version"
    const versionText = screen.getByText(
      new RegExp(translations.login_version_text, "i")
    ); // Case-insensitive match for version
    expect(versionText).toBeTruthy();

    // Check if the logo exists
    const logo = screen.getByTestId("ibe-logo");
    expect(logo).toBeTruthy();
  });

  /**
   * Test Case 2: Toggles password visibility
   * Verifies that the password visibility can be toggled using a button.
   */
  it("should toggle password visibility", () => {
    renderLogin();

    const passwordInput = screen.getByPlaceholderText(
      translations.login_password_placeholder
    );
    const toggleButton = screen.getByLabelText(translations.show_password);

    // Initially, the password should be hidden
    expect(passwordInput.props.secureTextEntry).toBe(true);

    fireEvent.press(toggleButton);

    // After pressing the toggle button, the password should be visible
    expect(passwordInput.props.secureTextEntry).toBe(false);

    // Verify the button label is updated correctly
    const updatedToggleButton = screen.getByLabelText(
      translations.hide_password
    ); // Check for updated label
    expect(updatedToggleButton).toBeTruthy();
  });

  /**
   * Test Case 3: Updates input values correctly
   * Verifies that the input fields for Username, Password, and Client ID update their values correctly.
   */
  it("should update input values correctly", () => {
    renderLogin();

    const usernameInput = screen.getByPlaceholderText(
      translations.login_username_placeholder
    );
    const passwordInput = screen.getByPlaceholderText(
      translations.login_password_placeholder
    );
    const clientIdInput = screen.getByPlaceholderText(
      translations.login_clientId_placeholder
    );

    fireEvent.changeText(usernameInput, "testuser");
    fireEvent.changeText(passwordInput, "testpass");
    fireEvent.changeText(clientIdInput, "123");

    // Assert the input values
    expect(screen.getByDisplayValue("testuser")).toBeTruthy();
    expect(screen.getByDisplayValue("testpass")).toBeTruthy();
    expect(screen.getByDisplayValue("123")).toBeTruthy();
  });

  /**
   * Test Case 4: Handles login successfully
   * Mocks a successful login request and checks if fetchData is called with the correct parameters.
   */
  it("should handle login successfully", async () => {
    fetchData.mockResolvedValueOnce({
      success: true,
      userid: "testuser",
    });

    renderLogin();

    const usernameInput = screen.getByPlaceholderText(
      translations.login_username_placeholder
    );
    const passwordInput = screen.getByPlaceholderText(
      translations.login_password_placeholder
    );
    const clientIdInput = screen.getByPlaceholderText(
      translations.login_clientId_placeholder
    );
    const loginButton = screen.getByRole("button", {
      name: translations.login_button_text,
    });

    // Simulate user input
    const username = "testuser";
    const password = "testpassword";
    const clientId = "123";

    fireEvent.changeText(usernameInput, username);
    fireEvent.changeText(passwordInput, password);
    fireEvent.changeText(clientIdInput, clientId);
    fireEvent.press(loginButton);

    await waitFor(() => {
      // Convert password to an array of ASCII values
      const passwordInASCII = Array.from(password, (char) =>
        char.charCodeAt(0)
      );

      // Prepare the expected payload with URL-encoded values for dynamic parts only
      const expectedPayload = `user=${encodeURIComponent(
        username
      )}&password=${encodeURIComponent(
        `[${passwordInASCII.join(",")}]`
      )}&enterpriseID=${encodeURIComponent(clientId)}`;

      // Verify the fetchData call
      expect(fetchData).toHaveBeenCalledWith(
        API_ENDPOINTS.AUTHENTICATE,
        "POST",
        expect.objectContaining({
          "Content-Type": "application/x-www-form-urlencoded; charset=UTF-8",
        }),
        expect.stringContaining(expectedPayload)
      );
    });
  });

  /**
   * Test Case 5: Shows error message when login fails
   * Mocks a failed login attempt and verifies that an error message is displayed.
   */
  it("should show error message when login fails", async () => {
    fetchData.mockResolvedValueOnce({
      success: false,
      param_msgText: translations.login_exception_message, // Use translated error message
    });

    renderLogin();

    const usernameInput = screen.getByPlaceholderText(
      translations.login_username_placeholder
    );
    const passwordInput = screen.getByPlaceholderText(
      translations.login_password_placeholder
    );
    const clientIdInput = screen.getByPlaceholderText(
      translations.login_clientId_placeholder
    );
    const loginButton = screen.getByRole("button", {
      name: translations.login_button_text,
    });

    fireEvent.changeText(usernameInput, "testuser");
    fireEvent.changeText(passwordInput, "testpass");
    fireEvent.changeText(clientIdInput, "123");
    fireEvent.press(loginButton);

    await waitFor(() => {
      expect(showToast).toHaveBeenCalledWith(
        translations.login_exception_message,
        "error"
      );
    });
  });

  /**
   * Test Case 6: Stores credentials when Remember Me is checked
   * Verifies that user credentials are stored securely when the Remember Me option is selected after a successful login.
   */
  it("should store credentials if Remember Me is checked", async () => {
    // Mock a successful API response
    fetchData.mockResolvedValueOnce({
      success: true,
    });

    renderLogin();

    // Get input elements and login button
    const usernameInput = screen.getByPlaceholderText(
      translations.login_username_placeholder
    );
    const passwordInput = screen.getByPlaceholderText(
      translations.login_password_placeholder
    );
    const clientIdInput = screen.getByPlaceholderText(
      translations.login_clientId_placeholder
    );
    const rememberMeCheckbox = screen.getByRole("checkbox", {
      name: translations.login_rememberMe_text,
    });

    const loginButton = screen.getByRole("button", {
      name: translations.login_button_text,
    });

    // Simulate user input
    fireEvent.changeText(usernameInput, "testuser");
    fireEvent.changeText(passwordInput, "testpass");
    fireEvent.changeText(clientIdInput, "123");
    fireEvent.press(rememberMeCheckbox); // Check the Remember Me checkbox
    fireEvent.press(loginButton);

    // Verify that checkbox is checked and credentials are stored securely
    await waitFor(() => {
      // Check if the checkbox is checked
      expect(rememberMeCheckbox).toBeChecked();

      expect(SecureStore.setItemAsync).toHaveBeenCalledWith(
        "username",
        "testuser"
      );
      expect(SecureStore.setItemAsync).toHaveBeenCalledWith("clientId", "123");
      expect(SecureStore.setItemAsync).toHaveBeenCalledWith(
        "password",
        "testpass"
      );
    });
  });

  /**
   * Test Case 7: Auto-login when "Remember Me" is checked
   * Verifies that the user is automatically logged in and the Home screen is displayed when
   * the "Remember Me" option was selected during the last login.
   */
  it("should auto-login user when credentials are found in SecureStore", async () => {
    // Mock the SecureStore.getItemAsync responses to simulate stored credentials
    SecureStore.getItemAsync
      .mockResolvedValueOnce("testuser") // Mock stored username
      .mockResolvedValueOnce("testpass") // Mock stored password
      .mockResolvedValueOnce("123"); // Mock stored clientId

    // Mock the fetchData function for a successful login response
    fetchData.mockResolvedValueOnce({
      success: true,
      userid: "testuser",
    });

    // Render the Login component
    renderLogin();

    // Verify that credentials are fetched from SecureStore
    await waitFor(() => {
      expect(SecureStore.getItemAsync).toHaveBeenCalledWith("username");
      expect(SecureStore.getItemAsync).toHaveBeenCalledWith("password");
      expect(SecureStore.getItemAsync).toHaveBeenCalledWith("clientId");
    });

    // Verify that the login API is called with the retrieved credentials
    await waitFor(() => {
      expect(fetchData).toHaveBeenCalledWith(
        expect.any(String), // endpoint string
        expect.any(String), // HTTP method string (e.g., "POST", "GET")
        expect.any(Object), // headers object
        expect.any(String) // body
      );
    });

    render(
      <MockNavigationContainer>
        <Home navigation={mockNavigation} />
      </MockNavigationContainer>
    );

    // Verify that the Home screen is displayed after auto-login
    const homeElement = screen.getByTestId("home-screen");
    expect(homeElement).toBeTruthy();
  });

  /**
   * Test Case 8: Shows error message when username is not provided
   * Verifies that an error message is shown if the username input field is empty.
   */
  it("should show error message when username is not provided", async () => {
    renderLogin();

    const passwordInput = screen.getByPlaceholderText(
      translations.login_password_placeholder
    );
    const clientIdInput = screen.getByPlaceholderText(
      translations.login_clientId_placeholder
    );
    const loginButton = screen.getByRole("button", {
      name: translations.login_button_text,
    });

    fireEvent.changeText(passwordInput, "testpass");
    fireEvent.changeText(clientIdInput, "123");
    fireEvent.press(loginButton);

    await waitFor(() => {
      expect(showToast).toHaveBeenCalledWith(
        translations.login_username_required,
        "error"
      );
    });
  });

  /**
   * Test Case 9: Shows error message when password is not provided
   * Verifies that an error message is shown if the password input field is empty.
   */
  it("should show error message when password is not provided", async () => {
    renderLogin();

    const usernameInput = screen.getByPlaceholderText(
      translations.login_username_placeholder
    );
    const clientIdInput = screen.getByPlaceholderText(
      translations.login_clientId_placeholder
    );
    const loginButton = screen.getByRole("button", {
      name: translations.login_button_text,
    });

    fireEvent.changeText(usernameInput, "testuser");
    fireEvent.changeText(clientIdInput, "123");
    fireEvent.press(loginButton);

    await waitFor(() => {
      expect(showToast).toHaveBeenCalledWith(
        translations.login_password_required,
        "error"
      );
    });
  });

  /**
   * Test Case 10: Shows error message when client ID is not provided
   * Verifies that an error message is shown if the client ID input field is empty.
   */
  it("should show error message when client ID is not provided", async () => {
    renderLogin();

    const usernameInput = screen.getByPlaceholderText(
      translations.login_username_placeholder
    );
    const passwordInput = screen.getByPlaceholderText(
      translations.login_password_placeholder
    );
    const loginButton = screen.getByRole("button", {
      name: translations.login_button_text,
    });

    fireEvent.changeText(usernameInput, "testuser");
    fireEvent.changeText(passwordInput, "testpass");
    fireEvent.press(loginButton);

    await waitFor(() => {
      expect(showToast).toHaveBeenCalledWith(
        translations.login_clientId_required,
        "error"
      );
    });
  });

  /**
   * Test Case 11: Shows error message for non-alphanumeric username
   * Ensures that an error message is shown when the username is not alphanumeric.
   */
  it("should show error message for non-alphanumeric username", async () => {
    renderLogin();

    const usernameInput = screen.getByPlaceholderText(
      translations.login_username_placeholder
    );
    const passwordInput = screen.getByPlaceholderText(
      translations.login_password_placeholder
    );
    const clientIdInput = screen.getByPlaceholderText(
      translations.login_clientId_placeholder
    );
    const loginButton = screen.getByRole("button", {
      name: translations.login_button_text,
    });

    fireEvent.changeText(usernameInput, "user@123");
    fireEvent.changeText(passwordInput, "testpass");
    fireEvent.changeText(clientIdInput, "123");
    fireEvent.press(loginButton);

    await waitFor(() => {
      expect(showToast).toHaveBeenCalledWith(
        translations.login_username_alphanumeric_required,
        "error"
      );
    });
  });

  /**
   * Test Case 12:
   * Test to verify that the error message is shown when username and password are incorrect.
   */
  it("should display error message when username and password are incorrect", async () => {
    // Mock API response for incorrect login credentials
    fetchData.mockResolvedValue({
      param_msgText: translations.login_invalid_credentials_message,
    });

    renderLogin();

    // Simulate user entering incorrect username, password, and client ID
    const usernameInput = screen.getByPlaceholderText(
      translations.login_username_placeholder
    );
    const passwordInput = screen.getByPlaceholderText(
      translations.login_password_placeholder
    );
    const clientIdInput = screen.getByPlaceholderText(
      translations.login_clientId_placeholder
    );

    fireEvent.changeText(usernameInput, "wrongUser");
    fireEvent.changeText(passwordInput, "wrongPass");
    fireEvent.changeText(clientIdInput, "123");

    // Simulate login button press
    const loginButton = screen.getByRole("button", {
      name: translations.login_button_text,
    });
    fireEvent.press(loginButton);

    // Wait for the login function to process and expect an error message to be shown
    await waitFor(() => {
      // Ensure the fetchData API call was made with the expected parameters
      expect(fetchData).toHaveBeenCalledWith(
        expect.any(String), // Expect any endpoint string
        expect.any(String), // Expect any HTTP method string (e.g., "POST", "GET")
        expect.any(Object), // Expect any headers object
        expect.any(String) // Expect any body string
      );

      // Verify that the toast is shown with the error message from the response
      expect(showToast).toHaveBeenCalledWith(
        translations.login_invalid_credentials_message,
        "error"
      );
    });
  });

  /**
   * Test Case 13: Snapshot test
   * Verifies that the rendered output matches the saved snapshot.
   */
  it("should match the snapshot", () => {
    const { toJSON } = renderLogin();
    expect(toJSON()).toMatchSnapshot(); // Take a snapshot of the rendered component
  });
});
