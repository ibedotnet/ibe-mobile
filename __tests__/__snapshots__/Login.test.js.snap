// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component should match the snapshot 1`] = `
[
  <View
    onLayout={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "flex": 1,
          "justifyContent": "center",
          "padding": "4%",
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
    testID="login-screen"
  >
    <View
      style={
        {
          "alignItems": "center",
          "flex": 2,
          "justifyContent": "center",
        }
      }
    >
      <Image
        accessibilityLabel="ibe logo"
        resizeMode="contain"
        source={1}
        style={
          {
            "width": 375,
          }
        }
        testID="ibe-logo"
      />
    </View>
    <View
      style={
        {
          "flex": 3,
          "justifyContent": "center",
          "width": "100%",
        }
      }
    >
      <TextInput
        accessibilityLabel="Username"
        maxLength={10}
        onChangeText={[Function]}
        placeholder="Username"
        placeholderTextColor="darkgrey"
        style={
          {
            "borderRadius": 28,
            "borderWidth": 2,
            "marginBottom": "4%",
            "paddingHorizontal": 16,
            "paddingVertical": 12,
          }
        }
        value=""
      />
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 28,
            "borderWidth": 2,
            "flexDirection": "row",
            "marginBottom": "4%",
            "paddingHorizontal": 16,
            "paddingVertical": 12,
            "width": "100%",
          }
        }
      >
        <TextInput
          accessibilityLabel="Password"
          autoCapitalize="none"
          autoCorrect={false}
          maxLength={20}
          onChangeText={[Function]}
          placeholder="Password"
          placeholderTextColor="darkgrey"
          secureTextEntry={true}
          style={
            {
              "flex": 1,
            }
          }
        />
        <View
          accessibilityLabel="Show password"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              [
                {
                  "color": "black",
                  "fontSize": 24,
                },
                undefined,
                {
                  "fontFamily": "ionicons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                {},
              ]
            }
          >
            î¯§
          </Text>
        </View>
      </View>
      <TextInput
        accessibilityLabel="Client ID"
        keyboardType="numeric"
        maxLength={4}
        onChangeText={[Function]}
        placeholder="Client ID"
        placeholderTextColor="darkgrey"
        style={
          {
            "borderRadius": 28,
            "borderWidth": 2,
            "marginBottom": "4%",
            "paddingHorizontal": 16,
            "paddingVertical": 12,
          }
        }
        value=""
      />
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "flex-end",
            "marginBottom": "4%",
          }
        }
      >
        <View
          accessibilityLabel="Remember Me"
          accessibilityRole="checkbox"
          accessibilityState={
            {
              "busy": undefined,
              "checked": false,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "borderColor": "#657786",
                "borderRadius": 2,
                "borderWidth": 2,
                "height": 20,
                "width": 20,
              },
              {
                "borderColor": "#000",
                "marginRight": 5,
              },
              false,
              false,
              undefined,
              false,
            ]
          }
        />
        <Text
          style={
            {
              "fontSize": 16,
            }
          }
        >
          Remember Me
        </Text>
      </View>
      <View
        accessibilityLabel="Login"
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#005eb8",
            "borderRadius": 30,
            "elevation": 5,
            "justifyContent": "center",
            "paddingVertical": "5%",
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.8,
            "shadowRadius": 2,
          }
        }
      >
        <Text
          style={
            {
              "color": "white",
              "fontSize": 16,
              "fontWeight": "bold",
              "letterSpacing": 0.5,
            }
          }
        >
          Login
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
          "width": "100%",
        }
      }
    >
      <Text>
        Version
         
        1.0.0
      </Text>
    </View>
  </View>,
  " ",
]
`;
